lint/src/main/java/leakcanary/lint/ProductionUsageDetector.kt

```diff
@@ -0,0 +1,54 @@
+package leakcanary.lint
+
+import com.android.tools.lint.detector.api.Category.Companion.CORRECTNESS
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.GradleContext
+import com.android.tools.lint.detector.api.Implementation
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.Scope
+import com.android.tools.lint.detector.api.Severity.ERROR
+import leakcanary.lint.ProductionUsageDetector.Companion.DETAIL_MESSAGE
+import leakcanary.lint.ProductionUsageDetector.Companion.ISSUE_ID
+import leakcanary.lint.ProductionUsageDetector.Companion.MESSAGE
+
+val ISSUE_PROD_USAGE_PATTERN = Issue.create(
+    ISSUE_ID,
+    MESSAGE,
+    DETAIL_MESSAGE,
+    CORRECTNESS, 10, ERROR,
+    Implementation(ProductionUsageDetector::class.java, Scope.GRADLE_SCOPE)
+)
+
+class ProductionUsageDetector : Detector(), Detector.GradleScanner {
+
+  companion object {
+    const val BLOCK_NAME = "dependencies"
+    const val DEPENDENCY_PREFIX = "com.squareup.leakcanary"
+    const val DESIRED_ENVIRONMENT = "debug"
+    const val MESSAGE = "LeakCanary should not be used in production."
+    const val DETAIL_MESSAGE = "$MESSAGE Please use debugImplementation instead."
+    const val ISSUE_ID = "ProductionUsagePattern"
+  }
+
+  override fun checkDslPropertyAssignment(
+    context: GradleContext,
+    property: String,
+    value: String,
+    parent: String,
+    parentParent: String?,
+    propertyCookie: Any,
+    valueCookie: Any,
+    statementCookie: Any
+  ) {
+    if (parent == BLOCK_NAME) {
+      if (value.contains(DEPENDENCY_PREFIX) &&
+          property.contains(DESIRED_ENVIRONMENT)
+              .not()
```

Comment:
It's a good point. I considered product flavors but not the build types. 

We can't just check for `property.contains("release")` as the `property` is either `implementation/api` or `debugImplementation/debugApi` but mostly not `release`.  I think we can assume that if the integrator uses `implementation/api` keywords to implement the library it will be affecting all the environments including `release`.  So, we can apply the lint rule if there is no prefix used or `release` keyword is included. 

`implementation 'com.squareup.leakcanary:leakcanary-android:x.y` -> Lint Error 
`releaseImplementation 'com.squareup.leakcanary:leakcanary-android:x.y` -> Lint Error 
`debugImplementation 'com.squareup.leakcanary:leakcanary-android:x.y` -> OK
`fullDebugImplementation 'com.squareup.leakcanary:leakcanary-android:x.y` -> OK

However, even with this approach there may be some cases we may not be able to catch. Say, there are `debug`, `release` and `staging` environments as well as `full` and `demo` build variants. With the approach above we would miss `fullImplementation` or `demoImplementation` usages as there is a prefix but it's not the environment itself. 

`fullImplementation 'com.squareup.leakcanary:leakcanary-android:x.y` -> Would miss this

I feel like we need to have the set of buildTypes and run the lint against them. But I don't think it'd be a straightforward implementation as the Lint API doesn't have the best documentation.

What do you think?

Reply:
IMO catching most of the cases would be sufficient, so your suggestions `apply the lint rule if there is no prefix used or release keyword is included` would work just right!

I wonder if Lint can even access list of all buildTypes/environments - seems like a too complicated task for it. 🤔

Reply:
Cool - I will update the PR to cover these cases then! Thanks for the review ✋